File: LinieLotnicze.ts
function WybierzTeKtoreZawieraja(lista, lotnisko){
    const ret = []
    for (let it in lista){
        if (it[0] == lotnisko || it[1] == lotnisko){
            ret.push(it)
        }
        return ret
    }
}

class Stan{
    poczatek : string;
    koniec : string;
    poprzednik : Stan;
    constructor ( poczatek = null, koniec = null, poprzednik = null) {
        this.poczatek = poczatek;
        this.koniec = koniec;
        this.poprzednik = poprzednik;
    }


}
var HistoriaStanow = [];
var StanyDoEkspansji = [];


function uaktualnij_stany(LS) {
    for (let s in LS){
        StanyDoEkspansji.push(s)
        HistoriaStanow.push(s)
    }
}

function jest_nowa(S) {
    for(const h of HistoriaStanow){
        if ((S.poczatek == h.poczatek && S.koniec == h.koniec) ||
            (S.koniec == h.poczatek && S.poczatek == h.koniec)){
            return false;
        }return true;
    }
}

function filtruj_stany(LS){
    var StanyOk = []
    for (let s in LS){
        if (jest_nowa(s)){
            StanyOk.push(s)
        } return StanyOk
    }
}



function czy_rozwiazanie(S, poczatek, koniec){
    if (S && S.poczatek == poczatek && S.koniec == koniec) {
        return true;
    }
    if (S.koniec == koniec){
        while (S){
            S = S.poprzednik
            if (S && S.poczatek == poczatek){
                return true;
            }
            else if (S && S.koniec == poczatek){
                return true;
            }
        }
    }
    if (S && S.koniec == poczatek){
        while (S){
            S = S.poprzednik
            if (S && S.poczatek == koniec){
                return true;
            }
            else if (S && S.koniec == koniec){
                return true;
            }
        }
        return false;
    }
}

function ekspanduj_stan(S, lista_polaczen) {
    var NoweStany = []
    for (let it in WybierzTeKtoreZawieraja(lista_polaczen, S.koniec)){
        if (S.koniec == it[0]) {
            NoweStany.push(new Stan(it[0], it[1], S));
        } else {
            NoweStany.push(new Stan(it[1], it[0], S));
        }
    }
    return NoweStany;
}

function wybierz_stan_do_ekspansji(){
    if ((StanyDoEkspansji.length) == 0){
        return null;
    }
    const S = StanyDoEkspansji[0];
    // @ts-ignore
    StanyDoEkspansji.pop(S);
    return S;
}



function main(pocztek, koniec, lista_polaczen){
    var poczatek_stanow = WybierzTeKtoreZawieraja(lista_polaczen, pocztek);
    for (let it in poczatek_stanow){
        if (pocztek == it[0]){
            S = new Stan(it[0], it[1]);
        }
        else{
            S = new Stan(it[1], it[0]);
            StanyDoEkspansji.push(S);
            HistoriaStanow.push(S);
        }
    }
    while (true){
        var S = wybierz_stan_do_ekspansji();
        if (S = null){
            console.log("Brak rozwiazania");
            break;
        }
        if (czy_rozwiazanie(S, pocztek, koniec)){
            var ret = []
            while (true){
                if (S && S.koniec){
                    ret.push((S.poczatek, S.koniec));
                    S = S.poprzednik;
                }else{
                    return ret;
                }

            }
            break;
        }

        var LS = ekspanduj_stan(S, lista_polaczen);
        LS = filtruj_stany(LS);
        uaktualnij_stany(LS);
    }
}

var lista_lotnisk = ['ATH', 'BSL', 'BFS', 'BLQ', 'BTS', 'BRS', 'CRL', 'BUD', 'DUB', 'EDI', 'EIN', 'GLA', 'HAM', 'CTA', 'KEF', 'CGN', 'SUF', 'LCA', 'LPL', 'LIS', 'LTN', 'STN', "MAD"]
var lista_polaczen = [['ATH', 'EDI'], ['ATH','GLA'], ['ATH','CTA'], ['BFS','CGN'], ['BFS','LTN'], ['BFS','CTA'],
    ['BTS', 'STN'], ['BTS', 'BLQ'], ['CRL','BLQ'], ['CRL','BSL'], ['CRL','LTN'], ['DUB','LCA'],
    ['LTN', 'DUB'], ['LTN',  'MAD'], ['LCA','HAM'], ['EIN','BUD'], ['EIN','MAD'], ['HAM','BRS'],
    ['KEF', 'LPL'], ['KEF', 'CGN'], ['SUF','LIS'], ['SUF','BUD'], ['SUF','STN'], ['STN','EIN'],
    ['STN', 'HAM'], ['STN', 'DUB'], ['STN','KEF']]

var lista;

lista = main('AEI', 'BRS', lista_polaczen)
lista.reverse()
for (let it in lista){
    console.log(it[0], it[1]);
}

File: LinieLotnicze.py
def WybierzTeKtoreZawieraja(lista, lotnisko):
    ret = []
    for it in lista:
        if it[0] == lotnisko or it[1] == lotnisko:
            ret.append(it)
    return ret


class Stan:
    def __init__(self, poczatek = None, koniec = None, poprzednik = None):
        self.poczatek = poczatek
        self.koniec = koniec
        self.poprzednik = poprzednik

    def __eq__(self, other):
        return (self.poczatek == other.poczatek and self.koniec == other.koniec) \
               or (self.koniec == other.poczatek and self.poczatek == other.koniec)


HistoriaStanow = []
StanyDoEkspansji = []


def uaktualnij_stany(LS):
    for s in LS:
        StanyDoEkspansji.append(s)
        HistoriaStanow.append(s)


def jest_nowa(S):
    for h in HistoriaStanow:
        if S == h:
            return False
    return True


def filtruj_stany(LS):
    StanyOk = []
    for s in LS:
        if jest_nowa(s):
            StanyOk.append(s)
    return StanyOk


def czy_rozwiazanie(S, poczatek, koniec):
    if S and S.poczatek == poczatek and S.koniec == koniec:
        return True
    if S.koniec == koniec:
        while S:
            S = S.poprzednik
            if S and S.poczatek == poczatek:
                return True
            elif S and S.koniec == poczatek:
                return True
    if S and S.koniec == poczatek:
        while S:
            S = S.poprzednik
            if S and S.poczatek == koniec:
                return True
            elif S and S.koniec == koniec:
                return True
    return False


def ekspanduj_stan(S, lista_polaczen):
    NoweStany = []
    for it in WybierzTeKtoreZawieraja(lista_polaczen, S.koniec):
        if S.koniec == it[0]:
            NoweStany.append(Stan(it[0], it[1], S))
        else:
            NoweStany.append(Stan(it[1], it[0], S))
    return NoweStany


def wybierz_stan_do_ekspansji():
    if len(StanyDoEkspansji) == 0:
        return None
    S = StanyDoEkspansji[0]
    StanyDoEkspansji.remove(S)
    return S


def main(pocztek, koniec, lista_polaczen):
    poczatek_stanow = WybierzTeKtoreZawieraja(lista_polaczen, pocztek)
    for it in poczatek_stanow:
        if pocztek == it[0]:
            S = Stan(it[0], it[1])
        else:
            S = Stan(it[1], it[0])
        StanyDoEkspansji.append(S)
        HistoriaStanow.append(S)

    while True:
        S = wybierz_stan_do_ekspansji()
        if S is None:
            print("Brak rozwiazanie")
            break
        if czy_rozwiazanie(S, pocztek, koniec):
            ret = []
            while True:
                if S and S.koniec:
                    ret.append((S.poczatek, S.koniec))
                    S = S.poprzednik
                else:
                    return ret
            break
        LS = ekspanduj_stan(S, lista_polaczen)
        LS = filtruj_stany(LS)
        uaktualnij_stany(LS)


lista_lotnisk = ['ATH', 'BSL', 'BFS', 'BLQ', 'BTS', 'BRS', 'CRL', 'BUD', 'DUB', 'EDI', 'EIN', 'GLA', 'HAM', 'CTA', 'KEF', 'CGN', 'SUF', 'LCA', 'LPL', 'LIS', 'LTN', 'STN', "MAD"]
lista_polaczen = [['ATH', 'EDI'], ['ATH','GLA'], ['ATH','CTA'], ['BFS','CGN'], ['BFS','LTN'], ['BFS','CTA'],
                  ['BTS', 'STN'], ['BTS', 'BLQ'], ['CRL','BLQ'], ['CRL','BSL'], ['CRL','LTN'], ['DUB','LCA'],
                  ['LTN', 'DUB'], ['LTN',  'MAD'], ['LCA','HAM'], ['EIN','BUD'], ['EIN','MAD'], ['HAM','BRS'],
                  ['KEF', 'LPL'], ['KEF', 'CGN'], ['SUF','LIS'], ['SUF','BUD'], ['SUF','STN'], ['STN','EIN'],
                  ['STN', 'HAM'], ['STN', 'DUB'], ['STN','KEF']]

print("Skąd chcesz odjechać?")
odjazd = input().upper()
print("Dokąd jedziesz?")
stacja_docelowa = input().upper()

if(odjazd not in lista_lotnisk or stacja_docelowa not in lista_lotnisk):
    print(" Nie ma takiego lotniska\n Spróbuj ponownie:")
    lista = main(odjazd, stacja_docelowa, lista_polaczen)

lista = main(odjazd, stacja_docelowa, lista_polaczen)
lista.reverse()
for it in lista:
    print(it[0], it[1])
